### DISCUSSION TITLE
Convert write process

### DISCUSSION THREAD

**Discussion by GRWilkie** [ImageMagick/ImageMagick#6416] [General]
Hi all,

I am a noob when it comes to imagemagick.   During a conversion process (from Tiff to PNG for example - convert image1.tif image1.png) can anyone confirm if the write process of the PNG file is done as a stream to the disk or does it in memory and then writes out as a single file?

I have a scenario where image files reside on a CIFS mounted share to the conversion server.  The conversion process takes around 8-10x as long compared to when the file is local to the conversion server.   Looking at network IO it's very consistent flat at 1-2Mb/sec through the conversion process suggesting its sending a stream of packets out to the disk (in this case the CIFS share).

Wondering if anyone has any experience of this sort of set up.
Thanks
Gordon

**Comment by snibgo** [comment]
If you use IM v7, I suggest using `magick` instead of `convert`. If your version of IM doesn't have `magick`, I suggest you upgrade it.

ImageMagick doesn't, itself, convert the image in memory to the new format, and then write that to disk.

The disk activity will vary by output format. IM uses delegate libraries for most formats. For example, when writing PNG, the IM coder module https://github.com/ImageMagick/ImageMagick/blob/main/coders/png.c , function WriteOnePNGImage(), does a load of preparation work which includes to writing to disk, and calls the libpng `png_write_row()` for each row in the image, then writes more stuff to the file.

You can see the real-time performance with:
```
magick -debug all image1.tif image1.png
```
This will list many operations to the screen (`stderr`), with timings.

You may get better overall performance if you write the image to a local disk (especially SSD), then copy that across the network using another program.

**Reply by GRWilkie** [reply]
Many thanks for this, I will ask our developers to look at this.
